standard/ranges.md

```diff
@@ -0,0 +1,309 @@
+  - a sequence is ***modifiable*** if every element can also be set using an *element_access* expression;
+  - an element’s from-start index is the number of elements before it in the sequence, for a sequence containing *N* elements:
+    - the first and last elements have indices of 0 and *N*-1 respectively, and
+    - the ***past-end index***, an index which represents a hypothetical element after the last one, has the value *N*.
+- A ***from-end index*** represents an element’s position within a sequence relative to the last element. For a sequence containing *N* elements the first, last and past-end indices are *N*, 1 and 0 respectively.
+- A ***range*** is a contiguous run of zero or more indices starting at any index within a sequence.
+- A ***slice*** is the collection of elements within a range.
+- A ***sliceable collection*** is one which:
+  - is countable;
+  - provides a method `Slice` which takes two `int` parameters specifying a range, being a starting index and a count of elements respectively, and returns a new slice constructed from the elements in the range.
+
+The above definitions are extended for uses of `Index` and `Range` as follows:
+
+- A type is also a *sequence* if an *element_access* expression taking a single required `Index` argument, rather than an `int` argument, is supported. Where a distinction is required the type is termed ***extended indexable***.
+- A type is also *sliceable* if an *element_access* expression taking a single required `Range` argument, rather a `Slice` method, is supported. Where a distinction is required the type is termed ***extended sliceable***.
+
+Whether a type is classified as countable, indexable, or sliceable is subject to the constraints of member accessibility (§7.5) and therefore dependent on where the type is being used.
+
+> *Example*: A type where the countable property and/or the indexer are `protected` is only a sequence to members of itself and any derived types. *end example*
+
+The required members for a type to qualify as a sequence or sliceable may be inherited.
+
+> *Example*: In the following code
+>
+> ```csharp
+> public class A
+> {
+>     public int Length { get { … } }
+> }
+>
+> public class B : A
+> {
+>     public int this(int index) { … }
+> }
+>
+> public class C : B
+> {
+>     public int[] Slice(int index, int count) { … }
+> }
+> ```
+>
+> The type `A` is countable, `B` is a sequence, and `C` is sliceable and a sequence.
+>
+> *end example*
+<!-- markdownlint-disable MD028 -->
+
+<!-- markdownlint-enable MD028 -->
+> *Note*:
+>
+> - A type can be sliceable without being indexable due to the lack of an (accessible) indexer.
+> - For a type to be sliceable and/or indexable requires the type to be countable.
+> - While the elements of a sequence are ordered by *position* within the sequence the elements themselves need not be ordered by their value, or even orderable.
+>
+> *end note*
+
+## 24.2 The Index type
```

Comment:
We should discuss these two sections.  

It's great work, but we have steadfastly avoided defining the semantics for library types. I see your point, but I'm concerned about increasing our workload when we are almost 6 releases behind.

Reply:
Personally I'm okay with defining the semantics of just Index and Range themselves, but not the semantics of how they're used.

Is the workload concern in terms of ongoing maintenance beyond the initial definition that has already been provided?

(No need to respond here - let's definitely discuss in the meeting.)

Reply:
I will argue we have no choice – the Standard needs to include the semantics either directly or by reference to another Standard, anything else makes the Standard incomplete. So we can discuss how we describe the semantics, have a goal of incrementally adding the semantics for the current contents of §C.3, require all new additions to §C.3 to come with semantics, etc. but we cannot say “semantics will not be provided”. C’est la vie.